django/test/testcases.py

```diff
@@ -730,6 +732,31 @@ def assertWarnsMessage(self, expected_warning, expected_message, *args, **kwargs
             *args, **kwargs
         )
 
+    # A similar method is available in Python 3.10+.
+    if not PY310:
+        @contextmanager
+        def assertNoLogs(self, logger_name, level=None):
+            """
+            Assert no messages are logged on the logger, with at least
+            the given level.
+
+            https://bugs.python.org/issue39385
+            """
+            if isinstance(level, int):
+                level = logging.getLevelName(level)
+            elif level is None:
+                level = 'INFO'
+            try:
+                with self.assertLogs(logger_name, level) as logs:
+                    yield
+            except AssertionError as e:
+                msg = e.args[0]
+                expected_msg = f'no logs of level {level} or higher triggered on {logger_name}'
+                if msg != expected_msg:
+                    raise e
+            else:
+                self.fail(f'logs of level {level} or higher triggered on {logger_name}:\n{logs.output!r}')
```

Comment:
I've proposed `self.fail(f'Unexpected logs found: {cm.output!r}')` to use the same message as in Python. With the current version `AssertNoLogsTest` tests will fail on Python 3.10+. We can adjust a message or skip new tests on Python 3.10+.

Reply:
I see, good idea. Thanks!